lib/Dialect/FIRRTL/Transforms/BlackboxMemory.cpp

```diff
@@ -0,0 +1,184 @@
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//===----------------------------------------------------------------------===//
+//
+// Transform memory operations in to instances of external modules for memory
+// generators.
+//
+//===----------------------------------------------------------------------===//
+
+#include "./PassDetails.h"
+#include "circt/Dialect/FIRRTL/FIRRTLOps.h"
+#include "circt/Dialect/FIRRTL/FIRRTLTypes.h"
+#include "circt/Dialect/FIRRTL/Passes.h"
+#include "llvm/ADT/Twine.h"
+
+using namespace circt;
+using namespace firrtl;
+
+static void replaceWithBlackbox(MemOp op) {
+  auto circuit = op->getParentOfType<CircuitOp>();
+  OpBuilder builder = OpBuilder::atBlockBegin(circuit.getBody());
+
+  // Get the name of the memory.  If none exists, use `mem`.
+  StringRef memName = "mem";
+  if (op.name().hasValue())
+    memName = op.name().getValue();
+  auto nameAttr = builder.getStringAttr(memName);
+
+  // Get the memory port descriptors.  This gives us the name and kind of each
+  // memory port created by the MemOp.
+  SmallVector<std::pair<Identifier, MemOp::PortKind>, 2> memPorts;
+  op.getPorts(memPorts);
+
+  // Create an external module for this blackbox memory. This external module
+  // must be compatible with the modules which are generated for `vlsi_mem_gen`,
+  // which can be found in the rocketchip project.
+  // The external module port list, 1R + 1W memory has 4+5=9 fields.
+  SmallVector<ModulePortInfo, 9> extPortList;
+  unsigned readPorts = 0;
+  unsigned writePorts = 0;
+  unsigned readWritePorts = 0;
+  for (unsigned i = 0; i < memPorts.size(); ++i) {
+    // Calculate the naming prefix to use based on the kind of the port.
+    std::string prefix;
+    switch (memPorts[i].second) {
+    case MemOp::PortKind::Read:
+      prefix = (Twine("R") + Twine(readPorts++) + "_").str();
+      break;
+    case MemOp::PortKind::Write:
+      prefix = (Twine("W") + Twine(writePorts++) + "_").str();
+      break;
+    case MemOp::PortKind::ReadWrite:
+      prefix = (Twine("RW") + Twine(readWritePorts++) + "_").str();
+      break;
+    }
+    // Flatten the bundle representing a memory port, name-mangling and adding
+    // every field in the bundle to the exter module's port list.
+    auto type = op.getResult(i).getType().cast<FIRRTLType>();
+    for (auto field : type.getPassiveType().cast<BundleType>().getElements()) {
+      auto fieldType = SubfieldOp::getResultType(type, field.name, op.getLoc());
+      extPortList.push_back({builder.getStringAttr(prefix + field.name.str()),
+                             FlipType::get(fieldType)});
+    }
+  }
+
+  // Create the blackbox external module.  The module's name is the name of the
+  // memory postfixed with "_ext".
+  std::string extName = memName.str() + "_ext";
+  auto extModule = builder.create<FExtModuleOp>(
+      op.getLoc(), builder.getStringAttr(extName), extPortList, extName);
```

Comment:
What if two memories have the same name?

Reply:
This has been fixed by using the SymbolTable. See test `NameCollision`.  I am guessing that the names generated here are nothing like the names generated by SCF.